File: assets/js/apps/applicationForm/models/vehicle.js
import Model from './model';

class Vehicle extends Model {}

Vehicle.fields = [
  {
    field: 'make_model',
    defaultValue: '',
    validation: (make_model) => make_model.length > 2 ? true : 'vehicle_mm_error'
  },
  {
    field: 'color',
    defaultValue: '',
    validation: (color) => color.length > 2 ? true : 'vehicle_color_error'
  },
  {
    field: 'license_plate',
    defaultValue: '',
    validation: (license_plate) => license_plate.length > 3  ? true : 'vehicle_license_error'
  },
  {
    field: 'state',
    defaultValue: '',
    validation: (state) => state.length === 2 ? true : 'vehicle_license_state_error'
  }
];

export default Vehicle;

File: lib/app_count/rent_apply/vehicle.ex
defmodule AppCount.RentApply.Vehicle do
  use Ecto.Schema
  import Ecto.Changeset
  @behaviour AppCount.RentApply.ValidatableBehaviour
  @derive {Poison.Encoder, only: [:color, :license_plate, :state, :make_model]}

  schema "rent_apply__vehicles" do
    field(:color, :string)
    field(:license_plate, :string)
    field(:state, :string)
    field(:make_model, :string)

    belongs_to(:application, Module.concat(["AppCount.RentApply.RentApplication"]),
      foreign_key: :application_id
    )

    timestamps()
  end

  @impl AppCount.RentApply.ValidatableBehaviour
  def validation_changeset(changeset, attrs) do
    changeset
    |> cast(attrs, [:make_model, :color, :license_plate, :state])
    |> validate_required([:make_model, :color, :license_plate, :state])
  end

  @doc false
  def changeset(%AppCount.RentApply.Vehicle{} = vehicle, attrs) do
    vehicle
    |> validation_changeset(attrs)
    |> cast(attrs, [:application_id])
    |> validate_required([:application_id])
  end
end
